docs/v2-recipes/developing-and-testing/debug-safe-app.md

```diff
@@ -0,0 +1,109 @@
+* Add breakpoints to it in your browser inspector.
+
+## I'm using VS Code
+VS Code allows "full stack" debugging i.e. both F# on the client and server. Prerequisites that you should install:
+
+#### 0. Install Prerequisites
+
+* **Install** either [Google Chrome](https://www.google.com/chrome/) or [Microsoft Edge](https://www.microsoft.com/en-us/edge): Enables client-side debugging.
+* **Configure your browser** with the following extensions:
+    * [Redux Dev Tools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en): Provides improved debugging support in Chrome with Elmish and access to Redux debugging.
+    * [React Developer Tools](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi/related): Provides access to React debugging in Chrome.
+* **Configure VS Code** with the following extensions:
+    * [Ionide](https://marketplace.visualstudio.com/items?itemName=Ionide.Ionide-fsharp): Provides F# support to Code.
+    * [C#](https://marketplace.visualstudio.com/items?itemName=ms-vscode.csharp): Provides .NET Core debugging support.
+    * [Debugger for Chrome](https://marketplace.visualstudio.com/items?itemName=msjsdiag.debugger-for-chrome): Provides integrated client-side debugging in Code.
+    * [Debugger for Edge](https://marketplace.visualstudio.com/items?itemName=msjsdiag.debugger-for-edge): Provides integrated client-side debugging in Code.
+
+#### 1. Create a launch.json file
+Open the Command Palette using `Ctrl+Shift+P` and run `Debug: Open launch.json`. This will ask you to choose a platform; select `.NET Core`.
+
+This will create a `launch.json` file in the root of your solution and also open it in the editor.
+
+#### 2. Add a Configuration
+Click the "Add Configuration..." button and choose **Launch .NET Core Console Application**.
+The only change required is to point it at the Server application, by replacing the `program` line with this:
+
+```json
+"program": "${workspaceFolder}/src/Server/bin/Debug/netcoreapp3.1/Server.dll",
+```
+
+#### 3. Configure a build task
+* From the Command Palette, choose `Configure Task`.
+* Select `Create tasks.json file from template`. This will show you a list of pre-configured templates.
+* Select `.NET Core`.
+
+#### 4. Debug the Server
+Either hit F5 or open the Debugging pane and press the Play button to build and launch the Server with the debugger attached.
+Observe that the Debug Console panel will show output from the server. The server is now running and you can set breakpoints and view the callstack etc.
+
+#### 5. Debug the Client
+* Start the Client using e.g. `npm run start`.
+* Open the Command Palette and run `Debug: Open Link`.
+* When prompted for a url, type `http://localhost:8080/`. This will launch a browser which is pointed at the URL and connect the debugger to it.
+* You can now set breakpoints in your Client within VS Code.
```

Comment:
As said here and we talked above, we can add the break points directly in the F# code in VS Code by using source maps. It's also possible to launch two VSCode debug processes (Server + Client) in one go using compounds. I've managed to "full-stack" debug my project with just one click using the following configuration (I'm omitting the Server config part because I'm using a node server).

tasks.json
```json
{
	"version": "2.0.0",
	"tasks": [
		{
			"label": "fable-watch",
			"isBackground": true,
			"type": "process",
			"command": "dotnet",
			"args": [
				"fable",
				"watch",
				"src/Client",
				"-s",
				"--run",
				"webpack-dev-server"
			]
		}
	]
}
```

launch.json
```json
{
    "version": "0.2.0",
    "compounds": [
        {
            "name": "Client+Server",
            "configurations": [ "Server", "Client" ]
        }
    ],
    "configurations": [
        {
            "type": "chrome",
            "request": "launch",
            "preLaunchTask": "fable-watch",
            "name": "Client",
            "url": "http://localhost:8080",
            "webRoot": "${workspaceFolder}/src",
            "sourceMaps": true,
            "sourceMapPathOverrides": {
                "webpack:///*": "${workspaceFolder}/*"
            }
        }
    ]
```

Reply:
Another v2 recipe

Reply:
Ah, ok. Then maybe the config can be used for the [comment above](https://github.com/SAFE-Stack/docs/pull/223#discussion_r654105492).